src/OpaqueVal.h

```diff
@@ -117,41 +138,61 @@ class OpaqueVal : public Val {
      * @return the broker representation, or an error if serialization
      * isn't supported or failed.
      */
-    std::optional<BrokerData> Serialize() const;
+    [[deprecated("use SerializeData instead")]] broker::expected<broker::data> Serialize() const;
+
+    /**
+     * @copydoc Serialize
+     */
+     std::optional<BrokerData> SerializeData() const;
 
     /**
      * Reinstantiates a value from its serialized Broker representation.
      *
      * @param data Broker representation as returned by *Serialize()*.
      * @return unserialized instances with reference count at +1
      */
-    static OpaqueValPtr Unserialize(BrokerDataView data);
+    [[deprecated("use UnserializeData instead")]] static OpaqueValPtr Unserialize(const broker::data& data);
+
+    /**
+     * @copydoc Unserialize
+     */
+    static OpaqueValPtr UnserializeData(BrokerDataView data);
 
     /**
      * @copydoc Unserialize
      */
-    static OpaqueValPtr Unserialize(BrokerListView data);
+    static OpaqueValPtr UnserializeData(BrokerListView data);
 
 protected:
     friend class Val;
     friend class OpaqueMgr;
 
+    /**
+     * @deprecated Override DoSerializeData instead.
+     */
+    virtual broker::expected<broker::data> DoSerialize() const;
```

Comment:
If this are marked with `[[deprecated]]`, will external plugins receive a deprecation warning when overriding it? If yes, we should do this and suppress the warnings for calling `DoSerialize()` in the default `DoSerializeData()` implementation in-tree (same for `DoUnserialize()`).

The doxygen `@deprecated` won't yield a compiler warning and so the casual user won't be notified explicitly.

Reply:
I had the same thought, but it seems like there's no way to deprecate overriding a virtual member function: https://stackoverflow.com/questions/49011990/c-deprecation-warning-when-overriding-a-deprecated-virtual-method. At least I don't know how we could get a warning for that. 🙁

I've added the `deprecated` Doxygen tag here to have at least *some* in-source documentation on that. Happy to hear better ideas.

Reply:
Oh, this is sad. For someone not reading NEWS this will pretty much end-up having the same effect as just breaking the interface, but I guess we can make the argument that we've documented it somewhere.

The other ideas I'm coming up revolve around adding some explicit info to derived class that they support the BrokerData serialization and minimally we could warn-once at runtime for those that don't. Probably not worth it.

Reply:
Suppose introducing a complete new `OpaqueVal` subclass (overridding the deprecated methods with final) and deprecating the current `OpaqueVal` could be a way out, but that would imply another name change and might be just too much.